aws-logging-cloudwatch/src/main/java/com/amplifyframework/logging/cloudwatch/AWSCloudWatchLoggingPluginImplementation.kt

```diff
@@ -0,0 +1,112 @@
+import com.amplifyframework.logging.cloudwatch.models.AWSCloudWatchLoggingPluginConfiguration
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+internal class AWSCloudWatchLoggingPluginImplementation(
+    private val loggingConstraintsResolver: LoggingConstraintsResolver,
+    private var awsCloudWatchLoggingPluginConfig: AWSCloudWatchLoggingPluginConfiguration? = null,
+    internal var cloudWatchLogManager: CloudWatchLogManager? = null,
+    dispatcher: CoroutineDispatcher = Dispatchers.IO,
+) {
+    private val defaultNameSpace = "amplify"
+    private val coroutineScope = CoroutineScope(dispatcher)
+    internal var isPluginEnabled = awsCloudWatchLoggingPluginConfig?.enable ?: true
+
+    internal fun configure(
+        awsPluginConfig: AWSCloudWatchLoggingPluginConfiguration,
+    ) {
+        awsCloudWatchLoggingPluginConfig = awsPluginConfig
+        loggingConstraintsResolver.localLoggingConstraint = awsPluginConfig.loggingConstraints
+        isPluginEnabled = awsPluginConfig.enable
+        if (isPluginEnabled) {
+            coroutineScope.launch {
+                cloudWatchLogManager?.startSync()
+            }
+        }
+        Amplify.Hub.subscribe(HubChannel.AUTH) { event ->
+            if (event.name == AuthChannelEventName.SIGNED_IN.name) {
+                cloudWatchLogManager?.onSignIn()
+            }
+            if (event.name == AuthChannelEventName.SIGNED_OUT.name) {
+                cloudWatchLogManager?.onSignOut()
+            }
+        }
+    }
+
+    internal fun forNamespace(namespace: String?): Logger {
+        val resolvedNameSpace = namespace ?: defaultNameSpace
+        return logger(resolvedNameSpace)
+    }
+
+    internal fun logger(namespace: String): Logger {
+        return CloudWatchLogger(
+            namespace,
+            null,
+            loggingConstraintsResolver,
+            this,
+        )
+    }
+
+    internal fun logger(categoryType: CategoryType, namespace: String): Logger {
+        return CloudWatchLogger(
+            namespace,
+            categoryType,
+            loggingConstraintsResolver,
+            this,
+        )
+    }
+
+    fun enable() {
+        isPluginEnabled = true
+        coroutineScope.launch {
+            cloudWatchLogManager?.startSync()
+        }
+    }
+
+    fun disable() {
+        isPluginEnabled = false
+        coroutineScope.launch {
+            cloudWatchLogManager?.stopSync()
+        }
+    }
+
+    internal fun flushLogs(
+        onSuccess: Action,
+        onError: Consumer<AmplifyException>,
+    ) {
+        coroutineScope.launch {
+            try {
+                cloudWatchLogManager?.syncLogEventsWithCloudwatch()
+                onSuccess.call()
+            } catch (exception: Exception) {
+                onError.accept(AmplifyException("Failed to flush logs", exception.cause, "Please try again"))
```

Comment:
Plugins should return `<Category>Exceptions` and not AmplifyExceptions (except from `configure()`).